src/shared/gpt.c

```diff
@@ -219,20 +211,36 @@ const char *gpt_partition_type_uuid_to_string_harder(
         return sd_id128_to_uuid_string(id, buffer);
 }
 
-int gpt_partition_type_from_string(const char *s, GptPartitionType *ret) {
+int gpt_partition_type_from_string(const char *s, Architecture arch, GptPartitionType *ret) {
         sd_id128_t id = SD_ID128_NULL;
+        PartitionDesignator d;
         int r;
 
         assert(s);
+        assert(arch >= 0);
+
+        /* We don't return immediately when we find a match, instead we re-resolve by UUID so that we can
+         * support aliases like aarch64 -> arm64 transparently. */
 
         for (size_t i = 0; i < ELEMENTSOF(gpt_partition_type_table) - 1; i++)
                 if (streq(s, gpt_partition_type_table[i].name)) {
-                        /* Don't return immediately, instead re-resolve by UUID so that we can support
-                        * aliases like aarch64 -> arm64 transparently. */
                         id = gpt_partition_type_table[i].uuid;
                         break;
                 }
 
+        /* If we can't find a matching name, fall back to parsing the name as a partition designator and
+         * matching against the partition designator + architecture. */
```

Comment:
honestly, i'd just leave gpt_partition_type_from_string() as it was before, and the gpt partition type table too. 

And then i'd implement gpt_partition_type_translate_arch() that takes a GptPartitionType structure as input and return the same one as output, but for the specified arch. and which searches the table based on target arch + designator, and returns the first hit.

i.e. code would then do:

```c
r = gpt_partition_type_from_string(s, &t);
if (r < 0) 
        return log_error(r, "Unknown partition type: %s", s);

if (arg_override_architecture >= 0) {
        r = gpt_partition_type_translate_architecture(t, arg_override_architecture, &t);
        if (r < 0)
                return log_error_errno(r, "Cannot translate partition type '%s' to arch '%s': %m", …);
 }
```

That's much nicer on the eye I think, and requires less patching around.

Now, one might argue that if somebody explicitly says "root-sparc64" instead of just "root" and then asks for translation to "arm" or so, then we should generate an error, but I am not sure if that's really necessary. If you insist though you could extend the table with another boolean "refuse_translation" or so which is set by _GPT_ARCH_SEXTET() for all arch-specific part types. but honestly I don't think anyone would be helped by that, hence i wouldn't bother.

Reply:
I tried this approach but it's a mess because we really only want --architecture to apply to the generic partition type names, not stuff like root-arm64 which should be arm64 regardless of --architecture I think.

I switched the approach so we don't need to remove anything from gpt_partition_type_table and handle secondary arches correctly as well, PTAL!

Reply:
as mentioned, i don't think we should necessarily generate an error when patching the architecture even for fully specified part types. It's a matter of documentation of me: if the docs just say something like this, I think we are good:

> --architecture= will write the image out for the specified architecture, replacing any specified partition type with the corresponding type for the selected architecture. For example, with `--architecture=arm64` a partition type of `root-x86-64` referenced in `repart.d/` drop-ins will be patched dynamically to refer to `root-arm64` instead.